engine/datahistory_manager.go

```diff
@@ -0,0 +1,870 @@
+	}
+
+	if job.DataType == dataHistoryTradeDataType &&
+		job.Interval <= 0 {
+		job.Interval = defaultTradeInterval
+	}
+
+	b := exch.GetBase()
+	if !b.Features.Enabled.Kline.Intervals[job.Interval.Word()] {
+		return fmt.Errorf("%s %w unsupported by exchange %s", job.Interval.Word(), kline.ErrUnsupportedInterval, job.Exchange)
+	}
+
+	job.StartDate = job.StartDate.Round(job.Interval.Duration())
+	job.EndDate = job.EndDate.Round(job.Interval.Duration())
+
+	return nil
+}
+
+// GetByID returns a job's details from its ID,
+func (m *DataHistoryManager) GetByID(id string) (*DataHistoryJob, error) {
+	if m == nil {
+		return nil, ErrNilSubsystem
+	}
+	if atomic.LoadInt32(&m.started) == 0 {
+		return nil, ErrSubSystemNotStarted
+	}
+	m.m.Lock()
+	for i := range m.jobs {
+		if m.jobs[i].ID.String() == id {
+			cpy := m.jobs[i]
+			m.m.Unlock()
+			return cpy, nil
+		}
+	}
+	m.m.Unlock()
+	dbJ, err := m.jobDB.GetByID(id)
+	if err != nil {
+		return nil, fmt.Errorf("%w with id %s %s", errJobNotFound, id, err)
+	}
+	result, err := m.convertDBModelToJob(dbJ)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert model with id %s %w", id, err)
+	}
+	return result, nil
+}
+
+// GetByNickname searches for jobs by name and returns it if found
+// returns nil if not
+// if fullDetails is enabled, it will retrieve all job history results from the database
+func (m *DataHistoryManager) GetByNickname(nickname string, fullDetails bool) (*DataHistoryJob, error) {
+	if m == nil {
+		return nil, ErrNilSubsystem
+	}
+	if atomic.LoadInt32(&m.started) == 0 {
+		return nil, ErrSubSystemNotStarted
+	}
+	if fullDetails {
+		dbJ, err := m.jobDB.GetJobAndAllResults(nickname)
+		if err != nil {
+			return nil, err
+		}
+		result, err := m.convertDBModelToJob(dbJ)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert model with nickname %s %w", nickname, err)
+		}
+		return result, nil
+	}
+	m.m.Lock()
+	for i := range m.jobs {
+		if strings.EqualFold(m.jobs[i].Nickname, nickname) {
+			cpy := m.jobs[i]
+			m.m.Unlock()
+			return cpy, nil
+		}
+	}
+	m.m.Unlock()
+	// now try the database
+	j, err := m.jobDB.GetByNickName(nickname)
+	if err != nil {
+		return nil, fmt.Errorf("%w, %s", errJobNotFound, err)
+	}
+	job, err := m.convertDBModelToJob(j)
+	if err != nil {
+		return nil, err
+	}
+
+	return job, nil
+}
+
+// GetAllJobStatusBetween will return all jobs between two ferns
+func (m *DataHistoryManager) GetAllJobStatusBetween(start, end time.Time) ([]*DataHistoryJob, error) {
+	dbJobs, err := m.jobDB.GetJobsBetween(start, end)
+	if err != nil {
+		return nil, err
+	}
+	var results []*DataHistoryJob
+	for i := range dbJobs {
+		dbJob, err := m.convertDBModelToJob(&dbJobs[i])
+		if err != nil {
+			return nil, err
+		}
+		results = append(results, dbJob)
+	}
+	return results, nil
+}
+
+// DeleteJob helper function to assist in setting a job to deleted
+func (m *DataHistoryManager) DeleteJob(nickname, id string) error {
```

Comment:
![image](https://user-images.githubusercontent.com/31364354/120761982-1cbf6a80-c559-11eb-85cc-cf681bb4a15e.png)
![image](https://user-images.githubusercontent.com/31364354/120762054-32349480-c559-11eb-9e09-d51af13856a4.png)
![Uploading image.png…]()

Reply:
Was trying to delete one of the jobs from GetActiveJobs using nickname and got the panic

3rd Imagine did not get inserted correctly 
![image](https://user-images.githubusercontent.com/31364354/120762202-598b6180-c559-11eb-9462-8c870de2554e.png)

Reply:
Sadly I cannot replicate this, could you provide any extra details? Can you still do it with any new jobs?

Reply:
We'll retest this soon